keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,944 @@
+other pods when scheduled on a node. Available policies are:
+
+-   `DaemonsetAndStaticPods`: pod only tolerates daemonset and static pods to be
+    running on the node.
+-   `Any`: pod tolerates any pod type to be running on the node (default)
+
+The API is implemented by the scheduler as a filter plugin and kubelet as an
+admission predicate.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants a Pod to be scheduled alone on a node to avoid them from contending
+on resources (e.g. claiming all CPU cores available on a node). In other words,
+resource intensive workloads are needed to be scheduled on separate nodes.
+
+Example: As a user, I want to deploy a Spark or MPI job such that only a single
+worker pod is running on each node.
+
+#### Story 2
+
+A DBaaS running natively on Kubernetes which uses 1:1 Pod to Node assignment
+where Nodes are sized appropriately for said pod's resource requirements. A
+combination of `nodeSelector`/`node affinity` with 1:1 pod to node assignment
+can be used to achieve this goal.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Users might be able to circumvent 1:1 pod to node assignments by creating a
+daemonset with node affinity to a specific node.
+
+#### Possible Mitigations
+
+-   Cluster admins can use one of the following to prevent users from creating
+    daemonsets
+    -   [ResourceQuota](https://kubernetes.io/docs/concepts/policy/resource-quotas/)
+        -   Object count quota (`count/daemonsets`) can be used for this
+            purpose.
+    -   kubernetes
+        [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) to
+        restrict daemonsets creation to admins/specific users.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The idea is to create a scheduler filter plugin such that given a pod and a
+node, a node is filtered out (as in considered no fit for the pod) if any of the
+following is true (assume `CoexistPolicy` is set to `DaemonsetAndStaticPods` in
+the incoming Pod's podspec):
+
+-   The incoming pod is not a daemonset pod or static pod (by checking
+    OwnerReference) and there is a pod with
```

Comment:
```suggestion
    OwnerReference) and there is a pod (that is not DaemonSet or Static Pod) with
```

Can static pod or daemonset has this set?

Reply:
Interesting! If we allow it, a DS pod with this policy would effectively repel any pods other than DS and static. This is similar to pod anti affinity with blank namespace and pod selector, so not really something new. I wouldn't explicitly disallow it (other than having a quota scope to limit which namespaces can use CoexistPolicy)

Reply:
I'd suggest to write it down explicitly so there is no assumption of the desired behavior

Reply:
+1. If this isn't supported it should probably also be a validation check on pod submit (for DS Pods)